### DISCUSSION TITLE
Request contains an invalid argument

### DISCUSSION THREAD

**Discussion by ppbittencourt1993** [kreait/firebase-php#918] [General]
Hi Jerome, how are you doing?

Some customers are presenting an issue when sending push notifications to their users's devices. 
The response when trying to send the message to some of the targeted devices is "Request contains an invalid argument", that doesn't give much to work with. For others is a success, using the same payload.
The message data seems to be created correctly, since otherwise it would fail for every device.

Another thing is that when this happens, the next push notification sent targeting the same devices, now returns a "Requested entity not found error" for each time the previous error happen, which apparently means that the error made the tokens be deleted from Firebase.

We were using version 5.x and recently updated to 7.10, but the issue happened while using both versions.

I keep investigating but i'm currently out of ideas.
Any thoughts on this?

**Comment by jeromegamez** [comment]
Hey Pablo,

I think this could have something to do with Firebase expiring tokens that have been inactive for more than 270 days.

> When stale tokens reach 270 days of inactivity, FCM will consider them expired tokens. Once a token expires, FCM marks it as invalid and rejects sends to it. However, FCM issues a new token for the app instance in the rare case that the device connects again and the app is opened. [Source](https://firebase.google.com/docs/cloud-messaging/manage-tokens#stale-and-expired-tokens)

This would explain why the first message reports an invalid field (the expired token), and after that the "Not found" error.

One way I could think of resolving this issue is to send a message with the `Messaging::sendAll()` method and the `validateOnly` parameter set to `true` for all tokens in your database. You can then catch all tokens that are reported as invalid/not found and delete them from your database.

If you're using the latest release, the `sendAll()` method should be fairly fast - if you have a ton of tokens, you could use a queue and multiple workers to speed things up.

Let me know if this helps!

**Reply by jeromegamez** [reply]
(Obligatory comment that I am looking for sponsors for my work on the SDK, especially when you're a company making serious money with it, as your mention of "some clients" suggests 😇)

https://github.com/sponsors/jeromegamez/

**Reply by ppbittencourt1993** [reply]
Hi Jerome, I really appreciate your work with the package and the quick replies.

Unfortunately it doesn't seem to be the case of the tokens expiring after 270 days of inactivity.
The device user subscribes to the channel and starts to receive notifications constantly for some days with a fresh token, but after some time (some days or weeks) he stops receiving the message, because apparently the error happened for his token.

Reading Firebase's documentation, I saw that the error 400 can occur for various reasons, like invalid parameters in the payload and message too big. I have forced those errors, to check the error message, but its very clear on what caused the error instead of the generic message I'm getting.

This is happening for a while now and its driving me crazy because I have no clue about what is causing it.

I will suggest for the company I work for that they sponsor your work, you definitely deserve and will try to sponsor you myself.

**Reply by jeromegamez** [reply]
Sad, I was hoping this would be it, it sounded really smart 😂

In general I have to say: if it works for one but not the other, then it's probably nothing I can fix in code. But I think that's what you think as well, otherwise you would have opened a bug report.

It does sound like a bug on the server side of FCM. You've done your research and probably have stumbled on the same SO questions as I have, with Firebase rolling back updates because something went wrong.

In the light of this I have to walk back from my suggestion to send validate-requests - if it is indeed a bug with the Firebase service, then even validate-only requests could result in tokens getting unregistered.

I read somewhere that someone had mistakenly sent messages to APNS Tokens instead of FCM tokens, but that can't be it either - if the tokens in questions were invalid, then the second request wouldn't result in a "Not found" error.

Hmmm...

Have you considered opening a support case with Firebase directly?

Can you share one of the messages in question (of course with sensitive data scrubbed, but with the same string lengths), so I can try it out with my one test-device and/or perhaps something in the message catches my eye.

Also, you mentioned that users are subscribed to a channel - you mean an FCM topic, yes? And what happened is that when you send the message to a topic, the device doesn't receive it, but when you send a message to the token, then you get this error?

When you have identified a token that doesn't receive messages via the topic anymore, could you try using the app instance API to get information about it? Perhaps, if it doesn't result in an invalid-argument error, it could give some insight...

https://firebase-php.readthedocs.io/en/7.13.1/cloud-messaging.html#app-instance-management

**Reply by jeromegamez** [reply]
Sharing the message is probably of not much use, after all you said that the same message goes through for other devices.

But I think it would be interesting if you could compare the registration tokens of working vs not working targets, just to be sure that they look similar.

**Reply by ppbittencourt1993** [reply]
Yeah, I don't believe its an issue with the package, feels more like a curse.

We are not using Firebase topics. We keep track of who we should send the messages to on our database.

This is an example of a message payload (sorry I had to hide most of it):
![image](https://github.com/user-attachments/assets/7f17dc71-a6f8-4847-9ed2-b808876ecb16)

I was able to get some tokens that were deleted from the database due to Firebase considering them invalid. They were valid yesterday.
OBS: I have removed 5 characters from them.

ef-g9dh4bkRZlrX9ZdeAs1:APA91bGkhK1zlIMNb0fvS-ams1TZHDPTdmEG66za0IXHDiH9P9hlH9A-4A50HTVRieaN8DkdhRvMbIzR8qjqyBk3gLcoqiyHGLOlWVwvASCmNSmhhS4RP7n1xoeVEnwwiG0yAst

eX9QJjyvx0VMtVIu8qwsd4:APA91bFSJRon0VcWEQExGM_sNI6vToBFuFVeBvHjOFH6X-77x-N7i1WLywvNqjON7oZKns9_Fc2qe2FdExjzsGWvBN1iqtZfTn9P79c-0drvxFO3ErCw9v8CzfgsvTG4wBHwpP_

cheEBjVyf08bvXcyhXJQ7H:APA91bGTfgPz3j7F5iITJvebFiej5J9QUNSdBLy-mPwoLB5YeEAnctiStdMuEOVqNIPRw0ulxQxwU5-c80RckaBJ_yB9dZfztlPgCf8Ozl51HLb_GhW87dtiZxJq-BFqalqtMRc

I could not spot anything wrong. They all have 163 characters like the valid tokens.

**Reply by jeromegamez** [reply]
I wish the SDK actually had a community, then perhaps someone could chime in, because I'm out of ideas.

I think the best way forward would be to open a support case with Firebase directly, as far as I know it's not paid support: 

There's also a Google Group at  https://groups.google.com/forum/#!forum/firebase-talk and a Slack Workspace at https://firebase.community/

**Reply by ppbittencourt1993** [reply]
Will do that,

Thank you very much for your assistance Jerome, you are awesome!

**Reply by jeromegamez** [reply]
You're very welcome! 🌺 Let me know if you find out more and/or if Firebase support can help you, I'm curious myself! If I come up with another idea, I'll let you know, too!

**Reply by ppbittencourt1993** [reply]
Will do that, thank you again!
